ocsp/responder/filter_source.go

```diff
@@ -0,0 +1,142 @@
+package responder
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/issuance"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+)
+
+type responderID struct {
+	nameHash []byte
+	keyHash  []byte
+}
+
+type filterSource struct {
+	wrapped        Source
+	hashAlgorithm  crypto.Hash
+	issuers        map[issuance.IssuerNameID]responderID
+	serialPrefixes []string
+	counter        *prometheus.CounterVec
+	log            blog.Logger
+}
+
+// NewFilterSource returns a filterSource which performs various checks on the
+// OCSP requests sent to the wrapped Source, and the OCSP responses returned
+// by it.
+func NewFilterSource(issuerCerts []*issuance.Certificate, serialPrefixes []string, wrapped Source, stats prometheus.Registerer, log blog.Logger) (*filterSource, error) {
```

Comment:
Go style would just call this `New`.

Reply:
Unfortunately I can't do that unless I push each of these filters down into their own subpackage, which I didn't deem to be worthwhile. I can totally do that if you'd prefer, though!

Reply:
That's fine!